libfwupd/fwupd-hsi-attr.c

```diff
@@ -0,0 +1,730 @@
+	if (flag == FWUPD_HSI_ATTR_FLAG_SUCCESS)
+		return "success";
+	if (flag == FWUPD_HSI_ATTR_FLAG_RUNTIME_UPDATES)
+		return "runtime-updates";
+	if (flag == FWUPD_HSI_ATTR_FLAG_RUNTIME_ATTESTATION)
+		return "runtime-attestation";
+	if (flag == FWUPD_HSI_ATTR_FLAG_RUNTIME_ISSUE)
+		return "runtime-issue";
+	if (flag == FWUPD_HSI_ATTR_FLAG_RUNTIME_UNTRUSTED)
+		return "runtime-untrusted";
+	return NULL;
+}
+
+/**
+ * fwupd_hsi_attr_get_obsoletes:
+ * @self: A #FwupdHsiAttr
+ *
+ * Gets the self obsoletes.
+ *
+ * Returns: (element-type utf8) (transfer none): the obsoletes, which may be empty
+ *
+ * Since: 1.4.2
+ **/
+GPtrArray *
+fwupd_hsi_attr_get_obsoletes (FwupdHsiAttr *self)
+{
+	FwupdHsiAttrPrivate *priv = GET_PRIVATE (self);
+	g_return_val_if_fail (FWUPD_IS_HSI_ATTR (self), NULL);
+	return priv->obsoletes;
+}
+
+/**
+ * fwupd_hsi_attr_add_obsolete:
+ * @self: A #FwupdHsiAttr
+ * @appstream_id: the attribute appstream_id
+ *
+ * Sets the attribute appstream_id.
+ *
+ * Since: 1.4.2
+ **/
+void
+fwupd_hsi_attr_add_obsolete (FwupdHsiAttr *self, const gchar *appstream_id)
+{
+	FwupdHsiAttrPrivate *priv = GET_PRIVATE (self);
+	g_return_if_fail (FWUPD_IS_HSI_ATTR (self));
+	g_return_if_fail (appstream_id != NULL);
+	for (guint i = 0; i < priv->obsoletes->len; i++) {
+		const gchar *obsolete_tmp = g_ptr_array_index (priv->obsoletes, i);
+		if (g_strcmp0 (obsolete_tmp, appstream_id) == 0)
+			return;
+	}
+	g_ptr_array_add (priv->obsoletes, g_strdup (appstream_id));
+}
+
+/**
+ * fwupd_hsi_attr_has_obsolete:
+ * @self: A #FwupdHsiAttr
+ * @appstream_id: the attribute appstream_id
+ *
+ * Finds out if the self has the attribute appstream_id.
+ *
+ * Returns: %TRUE if the self matches
+ *
+ * Since: 1.4.2
+ **/
+gboolean
+fwupd_hsi_attr_has_obsolete (FwupdHsiAttr *self, const gchar *appstream_id)
+{
+	FwupdHsiAttrPrivate *priv = GET_PRIVATE (self);
+	g_return_val_if_fail (FWUPD_IS_HSI_ATTR (self), FALSE);
+	g_return_val_if_fail (appstream_id != NULL, FALSE);
+	for (guint i = 0; i < priv->obsoletes->len; i++) {
+		const gchar *obsolete_tmp = g_ptr_array_index (priv->obsoletes, i);
+		if (g_strcmp0 (obsolete_tmp, appstream_id) == 0)
+			return TRUE;
+	}
+	return FALSE;
+}
+
+/**
+ * fwupd_hsi_attr_get_uri:
+ * @self: A #FwupdHsiAttr
+ *
+ * Gets the attribute uri.
+ *
+ * Returns: the attribute uri, or %NULL if unset
+ *
+ * Since: 1.4.2
+ **/
+const gchar *
+fwupd_hsi_attr_get_uri (FwupdHsiAttr *self)
+{
+	FwupdHsiAttrPrivate *priv = GET_PRIVATE (self);
+	g_return_val_if_fail (FWUPD_IS_HSI_ATTR (self), NULL);
+	return priv->uri;
+}
+
+/**
+ * fwupd_hsi_attr_set_uri:
+ * @self: A #FwupdHsiAttr
+ * @uri: the attribute URI
+ *
+ * Sets the attribute uri, i.e. where you can download the firmware from.
+ *
+ * Since: 1.4.2
+ **/
+void
+fwupd_hsi_attr_set_uri (FwupdHsiAttr *self, const gchar *uri)
+{
+	FwupdHsiAttrPrivate *priv = GET_PRIVATE (self);
+	g_return_if_fail (FWUPD_IS_HSI_ATTR (self));
+	g_free (priv->uri);
+	priv->uri = g_strdup (uri);
+}
```

Comment:
URI?  What is URI stuff doing here?  I didn't see that in design doc before

Reply:
Nuked.